pkg/bpfassets/attacher/libbpf_attacher.go

```diff
@@ -0,0 +1,334 @@
+				next = iterator.Next()
+				retry = 0
+			}
+			continue
+		}
+		getErr = binary.Read(bytes.NewBuffer(data), ByteOrder, &ct)
+		if getErr != nil {
+			klog.Infof("failed to decode received data: %v\n", getErr)
+			next = iterator.Next()
+			retry = 0
+			continue
+		}
+		if retry > 0 {
+			klog.V(5).Infof("successfully get data with retry=%d \n", retry)
+		}
+		processesData = append(processesData, ct)
+		keys = append(keys, key)
+		next = iterator.Next()
+		retry = 0
+	}
+	for _, key := range keys {
+		processes.DeleteKey(key)
+	}
+	return
+}
+
+func libbpfCollectFreq() (cpuFreqData map[int32]uint64, err error) {
+	cpuFreqData = make(map[int32]uint64)
+	var cpuFreq *bpf.BPFMap
+	cpuFreq, err = libbpfModule.GetMap(TableCPUFreqName)
+	if err != nil {
+		return
+	}
+	cpuFreqkeySize := int(unsafe.Sizeof(uint32Key))
+	iterator := cpuFreq.Iterator(cpuFreqkeySize)
+	var freq uint32
+	valueSize := int(unsafe.Sizeof(freq))
+	retry := 0
+	next := iterator.Next()
+	for next {
+		keyBytes := iterator.Key()
+		cpu := int32(ByteOrder.Uint32(keyBytes))
+		data, getErr := cpuFreq.GetValue(cpu, valueSize)
+		if getErr != nil {
+			retry += 1
+			if retry > maxRetry {
+				klog.Infof("failed to get data: %v with max retry: %d \n", getErr, maxRetry)
+				next = iterator.Next()
+				retry = 0
+			}
+			continue
+		}
+		getErr = binary.Read(bytes.NewBuffer(data), ByteOrder, &freq)
+		if getErr != nil {
+			klog.Infof("failed to decode received data: %v\n", getErr)
+			next = iterator.Next()
+			retry = 0
+			continue
+		}
+		if retry > 0 {
+			klog.V(5).Infof("successfully get data with retry=%d \n", retry)
+		}
+		cpuFreqData[cpu] = uint64(freq)
+		next = iterator.Next()
+		retry = 0
+	}
+	return
+}
+
+///////////////////////////////////////////////////////////////////////////
+// utility functions
+
+func unixOpenPerfEvent(bpfMap *bpf.BPFMap, typ, conf int) error {
+	perfKey := fmt.Sprintf("%d:%d", typ, conf)
+	sysAttr := &unix.PerfEventAttr{
+		Type:   uint32(typ),
+		Size:   uint32(unsafe.Sizeof(unix.PerfEventAttr{})),
+		Config: uint64(conf),
+	}
+
+	if _, ok := PerfEvents[perfKey]; ok {
+		return nil
+	}
+
+	cpus, err := getOnlineCPUs()
+	if err != nil {
+		return fmt.Errorf("failed to determine online cpus: %v", err)
+	}
+
+	res := []int{}
+
+	for _, i := range cpus {
+		cloexecFlags := unix.PERF_FLAG_FD_CLOEXEC
+
+		fd, err := unix.PerfEventOpen(sysAttr, -1, int(i), -1, cloexecFlags)
+		if fd < 0 {
+			return fmt.Errorf("failed to open bpf perf event: %v", err)
+		}
+		err = bpfMap.Update(int32(i), uint32(fd))
+		if err != nil {
+			return fmt.Errorf("failed to update bpf map: %v", err)
+		}
+		res = append(res, int(fd))
+	}
+
+	PerfEvents[perfKey] = res
+
+	return nil
+}
+
+func unixClosePerfEvent() {
+	for _, vs := range PerfEvents {
+		for _, fd := range vs {
+			unix.SetNonblock(fd, true)
+			unix.Close(fd)
+		}
+	}
+	PerfEvents = map[string][]int{}
+}
+
+func getOnlineCPUs() ([]uint, error) {
```

Comment:
what about reusing [this](https://github.com/sustainable-computing-io/kepler/blob/main/pkg/bpfassets/attacher/bcc_attacher.go#L127) function?